internal/http.go

```diff
@@ -0,0 +1,202 @@
+	ErrorLog *log.Logger
+}
+
+// HTTPError defines the http error codes
+type HTTPError struct {
+	code    int
+	message string
+	err     error
+}
+
+// errorResponse defines error response in json
+type errorResponse struct {
+	Error string `json:"error"`
+}
+
+func NewHTTPServer(ctx context.Context, httpServerPort string, conn *pgx.Conn, logger logr.Logger) {
+	http.HandleFunc("/download", func(w http.ResponseWriter, r *http.Request) {
+		httpErr := downloadReport(ctx, w, r, conn, logger)
+		if httpErr != nil {
+			writeErrorResponse(logger, w, httpErr.code, httpErr.message)
+		}
+	})
+}
+
+func downloadReport(ctx context.Context, w http.ResponseWriter, r *http.Request, conn *pgx.Conn, logger logr.Logger) *HTTPError {
+	HTTPErr := validateRequest(r)
+	if HTTPErr != nil {
+		return &HTTPError{HTTPErr.code, HTTPErr.message, nil}
+	}
+
+	err := generateReport(ctx, conn, r, logger)
+	if err != nil {
+		return &HTTPError{http.StatusInternalServerError, "error generating report", err}
+	}
+
+	w.WriteHeader(http.StatusOK)
+
+	_, err = w.Write([]byte("generated report"))
+	if err != nil {
+		logger.Error(err, "failed writing http response:")
+	}
+
+	return nil
+}
+
+func validateRequest(r *http.Request) *HTTPError {
+	if r.URL.Path != "/download" {
+		return &HTTPError{http.StatusNotFound, "404 not found", nil}
+	}
+
+	if r.Method != "GET" {
+		return &HTTPError{http.StatusMethodNotAllowed, "method is not supported", nil}
+	}
+
+	query := r.URL.Query()
+	start, present := query["start"]
+	if !present || len(start) == 0 {
+		return &HTTPError{http.StatusBadRequest, "start time not present", nil}
+	}
+
+	end, present := query["end"]
+	if !present || len(end) == 0 {
+		return &HTTPError{http.StatusBadRequest, "end time not present", nil}
+	}
+	return nil
+}
+
+func generateReport(ctx context.Context, conn *pgx.Conn, r *http.Request, logger logr.Logger) error {
+	query := r.URL.Query()
+	start := query["start"]
+	end := query["end"]
+	dirName := "/tmp/curator-report"
+	err := os.MkdirAll(dirName, 0755)
+	if err != nil {
+		logger.Error(err, "error making dir:")
+		return err
+	}
+	tarDirName := fmt.Sprintf("%s/%s-%s-koku-metrics.tar.gz",
+		dirName, strings.Join(start, ","), strings.Join(end, ","))
+	outputFile, err := os.OpenFile(tarDirName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		logger.Error(err, "error opening file:")
+		return err
+	}
+	defer outputFile.Close()
+	writer, err := gzip.NewWriterLevel(outputFile, gzip.BestCompression)
+	if err != nil {
+		logger.Error(err, "error:")
+		return err
+	}
+	defer writer.Close()
+	tw := tar.NewWriter(writer)
+	defer tw.Close()
+	err = exporter(ctx, conn, tw, r, logger)
+	if err != nil {
+		return err
+	}
+	logger.Info("report generated successfully")
+	return nil
+}
+
+func exporter(ctx context.Context, conn *pgx.Conn, tw *tar.Writer, r *http.Request, logger logr.Logger) error {
+	query := r.URL.Query()
+	start := query["start"]
+	end := query["end"]
+
+	for i := 0; i < 4; i++ {
```

Comment:
not blocking/follow-up: what's significance of 4 here? [Magic number](https://en.wikipedia.org/wiki/Magic_number_(programming))?